cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -5039,6 +5055,114 @@ if(ARROW_S3)
   endif()
 endif()
 
+# ----------------------------------------------------------------------
+# Azure SDK for C++
+
+macro(build_azure_sdk)
+  message(STATUS "Building Azure SDK for C++ from source")
+
+  find_curl()
+  find_package(LibXml2 REQUIRED)
+  add_custom_target(azure_sdk_dependencies)
+
+  set(AZURE_SDK_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/azure-sdk_ep-install")
+  set(AZURE_SDK_INCLUDE_DIR "${AZURE_SDK_PREFIX}/include")
+
+  set(AZURE_SDK_COMMON_CMAKE_ARGS
+      ${EP_COMMON_CMAKE_ARGS}
+      "-DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>"
+      "-DCMAKE_PREFIX_PATH=${AZURE_SDK_PREFIX}"
+      -DDISABLE_AZURE_CORE_OPENTELEMETRY=ON
+      -DWARNINGS_AS_ERRORS=OFF 
+      -DVCPKG_OVERLAY_PORTS=${CMAKE_SOURCE_DIR}/overlays)
+
+  file(MAKE_DIRECTORY ${AZURE_SDK_INCLUDE_DIR})
+  set(AZURE_CORE_STATIC_LIBRARY
+      "${AZURE_SDK_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}azure-core${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  set(AZURE_IDENTITY_STATIC_LIBRARY
+      "${AZURE_SDK_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}azure-identity${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  set(AZURE_STORAGE_BLOBS_STATIC_LIBRARY
+      "${AZURE_SDK_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}azure-storage-blobs${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  set(AZURE_STORAGE_COMMON_STATIC_LIBRARY
+      "${AZURE_SDK_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}azure-storage-common${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  set(AZURE_STORAGE_FILES_DATALAKE_STATIC_LIBRARY
+      "${AZURE_SDK_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}azure-storage-files-datalake${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  externalproject_add(azure-sdk_ep
+                      ${EP_COMMON_OPTIONS}
+                      INSTALL_DIR ${AZURE_SDK_PREFIX}
+                      URL ${ARROW_AZURE_SDK_URL}
+                      URL_HASH "SHA256=${ARROW_AZURE_SDK_BUILD_SHA256_CHECKSUM}"
+                      CMAKE_ARGS ${AZURE_SDK_COMMON_CMAKE_ARGS}
+                      BUILD_BYPRODUCTS ${AZURE_CORE_STATIC_LIBRARY}
+                                       ${AZURE_IDENTITY_STATIC_LIBRARY}
+                                       ${AZURE_STORAGE_BLOBS_STATIC_LIBRARY}
+                                       ${AZURE_STORAGE_COMMON_STATIC_LIBRARY}
+                                       ${AZURE_STORAGE_FILES_DATALAKE_STATIC_LIBRARY}
+                      DEPENDS azure_sdk_dependencies)
+  add_library(Azure::azure-core STATIC IMPORTED)
+  set_target_properties(Azure::azure-core
+                        PROPERTIES IMPORTED_LOCATION ${AZURE_CORE_STATIC_LIBRARY}
+                                   INTERFACE_INCLUDE_DIRECTORIES
+                                   ${AZURE_SDK_INCLUDE_DIR})
+  set_property(TARGET Azure::azure-core
+               PROPERTY INTERFACE_LINK_LIBRARIES CURL::libcurl OpenSSL::SSL)
+  add_dependencies(Azure::azure-core azure-sdk_ep)
```

Comment:
It seems to work fine without :+1:

Reply:
I tested this. It worked on the Ubuntu 22 build but I think on Ubuntu 20 the version of curl is too old. 
```
/home/tomnewton/arrow/cpp/build/_deps/azure_sdk-src/sdk/core/azure-core/src/http/curl/curl.cpp: In constructor ‘Azure::Core::Http::CurlConnection::CurlConnection(Azure::Core::Http::Request&, const Azure::Core::Http::CurlTransportOptions&, const string&, const string&)’:
/home/tomnewton/arrow/cpp/build/_deps/azure_sdk-src/sdk/core/azure-core/src/http/curl/curl.cpp:2313:5: error: ‘curl_blob’ was not declared in this scope
 2313 |     curl_blob rootCertBlob
      |     ^~~~~~~~~
/home/tomnewton/arrow/cpp/build/_deps/azure_sdk-src/sdk/core/azure-core/src/http/curl/curl.cpp:2318:37: error: ‘CURLOPT_CAINFO_BLOB’ was not declared in this scope; did you mean ‘CURLOPT_CAINFO’?
 2318 |     if (!SetLibcurlOption(m_handle, CURLOPT_CAINFO_BLOB, &rootCertBlob, &result))
      |                                     ^~~~~~~~~~~~~~~~~~~
      |                                     CURLOPT_CAINFO
/home/tomnewton/arrow/cpp/build/_deps/azure_sdk-src/sdk/core/azure-core/src/http/curl/curl.cpp:2318:59: error: ‘rootCertBlob’ was not declared in this scope
 2318 |     if (!SetLibcurlOption(m_handle, CURLOPT_CAINFO_BLOB, &rootCertBlob, &result))
      |          
```
I think previously it was fetching curl from vcpkg. With this change it seems to be using the system one. Probably that is what we would want except the Azure SDK appears to be incompatible with older versions.

Reply:
I can see 2 options:
1. We allow some vcpkg stuff that fetches a new enough version of curl. 
2. We use `BUILD_TRANSPORT_CURL=OFF`. I believe this removes the dependency on curl. However I think it prevents us from building the whole SDK. II've been struggling a bit to build just the required elements with the FetchContent version.

Reply:
3. Can we require recent curl for `ARROW_AZURE=ON`? (We don't support `ARROW_AZURE=ON` on Ubuntu 20.04.)

I think that mixing system curl and vcpkg's curl in the same process will cause a problem...

Reply:
This is done apart from I wanted to added some kind of explicit requirement on curl `>=7.71.0`

Reply:
It seems that this was fixed by https://github.com/Azure/azure-sdk-for-cpp/pull/4821 .
How about using more recent Azure SDK for C++?

Reply:
Ah, Ubuntu 20.04's curl is 7.68.0...

Reply:
This is done now. We're using `azure-core_1.10.2` since that is the latest non-beta release referencing a component we use. We need to use a release for a component we use in order to apply the same version constraint in `FindAzure.cmake`.

I've also re-enable the ubuntu 20 build since that is now working,